pkg/util/tls_certificate.go

```diff
@@ -0,0 +1,118 @@
+package util
+
+import (
+	"bytes"
+	"crypto/tls"
+	"os"
+	"sync"
+	"time"
+
+	pb "github.com/buildbarn/bb-storage/pkg/proto/configuration/tls"
+)
+
+// RotatingTLSCertificate provides an up-to-date certificate given file paths and a refresh interval.
+type RotatingTLSCertificate struct {
+	sync.RWMutex
+	acceptCert  func(tls *tls.Certificate) error
+	cert        *tls.Certificate
+	certData    []byte
+	certFile    string
+	keyData     []byte
+	keyFile     string
+	refreshStop chan bool
+}
+
+// NewRotatingTLSCertificate creates TLS certificate provider
+// object based on parameters specified in a Protobuf message for use
+// with a TLS client. This Protobuf message is embedded in Buildbarn
+// configuration files.
+func NewRotatingTLSCertificate(configuration *pb.X509KeyPair_Files, acceptCert func(*tls.Certificate) error) (*RotatingTLSCertificate, error) {
+	r := &RotatingTLSCertificate{
+		certFile:    configuration.CertificatePath,
+		keyFile:     configuration.PrivateKeyPath,
+		refreshStop: make(chan bool),
+		acceptCert:  acceptCert,
+	}
+
+	if err := configuration.RefreshInterval.CheckValid(); err != nil {
+		return nil, StatusWrap(err, "Failed to parse refresh interval")
+	}
+
+	if err := r.loadCertificate(); err != nil {
+		return nil, err
+	}
+
+	// Start refreshing the certificate on the desired interval.
+	go r.refreshOnInterval(configuration.RefreshInterval.AsDuration())
+
+	return r, nil
+}
+
+// GetCertificate provides the most recently obtained semantically valid tls.Certificate.
+func (r *RotatingTLSCertificate) GetCertificate() *tls.Certificate {
+	r.RLock()
+	defer r.RUnlock()
+	return r.cert
+}
+
+func (r *RotatingTLSCertificate) loadCertificate() error {
+	// Read certificate data from disk.
+	certData, err := os.ReadFile(r.certFile)
+	if err != nil {
+		return StatusWrap(err, "Failed to read certificate file")
+	}
+	keyData, err := os.ReadFile(r.keyFile)
+	if err != nil {
+		return StatusWrap(err, "Failed to read private key file")
+	}
+
+	// Read the existing cert data, to see if its changed.
+	r.RLock()
+	sameCredentials := bytes.Equal(r.certData, certData) && bytes.Equal(r.keyData, keyData)
+	r.RUnlock()
+
+	// Skip updating if the certificate has not changed.
+	if sameCredentials {
+		return nil
+	}
+
+	// Parse the PEM data into a certificate.
+	cert, err := tls.X509KeyPair(certData, keyData)
+	if err != nil {
+		return StatusWrap(err, "Invalid certificate file or private key file")
+	}
+
+	if err := r.acceptCert(&cert); err != nil {
+		return nil
+	}
+
+	// Update the certificate.
+	r.Lock()
+	r.cert = &cert
+	r.certData = certData
+	r.keyData = keyData
+	r.Unlock()
+
+	return nil
+}
+
+// StopRefresh stops reloading the certificate from files.
+func (r *RotatingTLSCertificate) StopRefresh() {
+	r.refreshStop <- true
+}
+
+func (r *RotatingTLSCertificate) refreshOnInterval(refreshInterval time.Duration) {
+	t := time.NewTicker(refreshInterval)
+	for {
+		select {
+		case <-r.refreshStop:
```

Comment:
My recommendation would be to remove the spawning of the goroutine from this type. Just make it so that there is a single `Refresh()` method on this, and let tls.go spawn the goroutine. That way the tests remain simple. We also won't need to care about the refreshing then, as tls.go can simply let it run indefinitely.

Reply:
Sounds good to me, that makes sense!